aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -44,9 +68,22 @@ struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct
 */
 void aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
 {
-    /**
-    * TODO: implement per description
-    */
+    if (!buffer->full) {
+        buffer->entry[buffer->in_offs] = *add_entry;
+        buffer->in_offs = (buffer->in_offs + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+        
+        // Check if the buffer is full or not
+        if (buffer->in_offs == buffer->out_offs) {
+            buffer->full = true;
+        }
+    } 
+    else 
+    {
+        // Overwrite the oldest entry and advance the out offset
+        buffer->entry[buffer->in_offs] = *add_entry;
+        buffer->in_offs = (buffer->in_offs + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
```

Comment:
@AshwiniCUB you have some duplicate logic in the if/else branches here which could be simplified.